packages/unified-selection/src/unified-selection/SelectionStorage.ts

```diff
@@ -0,0 +1,268 @@
+   * Get the selection levels currently stored for the specified imodel
+   * @param imodel iModel connection to get selection levels for.
+   * */
+  getSelectionLevels(imodel: IModelConnection): number[];
+  /** Get the selection stored in the storage.
+   * @param imodel iModel connection which the selection is associated with.
+   * @param level Level of the selection
+   */
+  getSelection(imodel: IModelConnection, level: number): SelectableObjectSet;
+  /**
+   * Add keys to the selection
+   * @param source Name of the selection source
+   * @param imodel iModel associated with the selection
+   * @param selectableObjects selectable objects to add
+   * @param level Selection level
+   * @param rulesetId ID of the ruleset in case the selection was changed from a rules-driven control
+   */
+  addToSelection(source: string, imodel: IModelConnection, selectableObjects: SelectableObjects, level: number, rulesetId?: string): void;
+  /**
+   * Remove keys from current selection
+   * @param source Name of the selection source
+   * @param imodel iModel associated with the selection
+   * @param selectableObjects selectableObjects to remove
+   * @param level Selection level
+   * @param rulesetId ID of the ruleset in case the selection was changed from a rules-driven control
+   */
+  removeFromSelection(source: string, imodel: IModelConnection, selectableObjects: SelectableObjects, level: number, rulesetId?: string): void;
+  /**
+   * Replace current selection
+   * @param source Name of the selection source
+   * @param imodel iModel associated with the selection
+   * @param selectableObjects selectableObjects to replace the current selection with
+   * @param level Selection level
+   * @param rulesetId ID of the ruleset in case the selection was changed from a rules-driven control
+   */
+  replaceSelection(source: string, imodel: IModelConnection, selectableObjects: SelectableObjects, level: number, rulesetId?: string): void;
+  /**
+   * Clear current selection
+   * @param source Name of the selection source
+   * @param imodel iModel associated with the selection
+   * @param level Selection level
+   * @param rulesetId ID of the ruleset in case the selection was changed from a rules-driven control
+   */
+  clearSelection(source: string, imodel: IModelConnection, level: number, rulesetId?: string): void;
+}
+
+/**
+ * The selection storage which stores the overall selection.
+ * @beta
+ */
+export class SelectionStorage implements SelectionStorageInterface {
+  private _storage = new Map<IModelConnection, SelectableObjectContainer>();
+
+  /** An event which gets broadcasted on selection changes */
+  public readonly selectionChange: StorageSelectionChangeEvent;
+
+  /**
+   * Creates an instance of `SelectionStorage`.
+   */
+  constructor() {
+    this.selectionChange = new StorageSelectionChangeEvent();
+    IModelConnection.onClose.addListener((imodel: IModelConnection) => {
+      this.onConnectionClose(imodel);
+    });
```

Comment:
Without a dependency on `IModelConnection` we won't be able to add this listener. Need to come up with a plan for handling cleanup.

Reply:
Now that I think of it, maybe it's enough to add some docs for `createStorage` explaining that the consumer needs to call `clearStorage` when iModel is closed.